### PR TITLE
WIP: Fix Markdown rendering issues

### PR DESCRIPTION
**Pull request by dregad**
I started working on a branch targeted at fixing the various rendering issues following the enabling of Parsedown safe mode ([#24186](https://www.mantisbt.org/bugs/view.php?id=24186)).

The work item covers refactoring of the following MantisCoreFormatting plugin features:
- [x] user mentions
- [x] URL and e-mail addresses
- [ ] Bug and Bugnote linking

This is still a work-in-progress at this time, but in the interest of getting some early testing and feedback, hopefully also by the people currently affected by the issue who are facing the difficult choice between disabling Markdown or leaving their MantisBT instance vulnerable to XSS, I thought I'd open a PR now.

Warning: this feature branch is subject to change (including rebasing) without notice.

Note: The PR currently includes a Composer update to a beta version of Parsedown, but it should also be possible to fix the issue (in a less elegant / complete way) with current 1.7.1 version. I'll decide which version to use later on (depending on when Parsedown 1.8.0 is released).

Fixes [#24628](https://www.mantisbt.org/bugs/view.php?id=24628), [#24241](https://www.mantisbt.org/bugs/view.php?id=24241)

### CODE REVIEW

File: core/mention_api.php [hunk 1]
```diff
@@ -73,7 +73,10 @@ function mention_get_candidates( $p_text ) {
 			# Negative lookbehind  to ensure we don't match multiple tags
 			. '(?<!' . $t_quoted_tag . ')' . $t_quoted_tag
 			. ')'
+			# This is the main username matching pattern:
 			# any word char, dash or period, must end with word char
+			# It must be consistent and kept in sync with the one defined in
+			# MantisMarkdown class
 			. '([\w\-.]*[\w])'
```
Comment (vboctor) [inline core/mention_api.php]: Should we have a constant or a function to enforce consistency of this expression?

File: plugins/MantisCoreFormatting/core/MantisMarkdown.php [hunk 2]
```diff
@@ -216,6 +230,105 @@ protected function inlineCode( $block ) {
 		return $block;
 	}
 
+	/**
+	 * Lookup the
```
Comment (vboctor) [inline plugins/MantisCoreFormatting/core/MantisMarkdown.php]: Documentation incomplete.

**Comment by dregad** [comment]
This is caused by MantisCoreFormatting - ``` ` ``` is a valid e-mail address char as per our regex, but not for the hostname part, which explains the positioning of the closing `>`. I'll see how this can be fixed.

**Comment by ajtruckle** [comment]
I am still not sure of the right way to test the new changes. All I have is my live mantis support site.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by atrol** [comment]
> What is the safest way for me to try this without blowing up my live ssystem on the website?

If you want the safest way, there can't be no other answer than to setup a separate test system.

**Comment by rimy** [comment]
@dregad and MantisBT team, thanks a lot for your work, we are using MantisBT with Markdown in production, it is an excellent bug tracking system, and I just did a fresh install to help testing this issue:

0. First of all, I am new to git and github, if I did anything wrong, please let me know, and I am happy to learn more. And I am not native English speaker, sorry for my English.

1. Environment
    Ubuntu 18.04.2 LTS
    git 2.17.1
    Apache/2.4.29
    PHP 7.2.17
    MariaDB 10.1.38

2. It seems does not works with current parsedown with 2.18.0, here is my steps:
    $ git clone https://github.com/dregad/mantisbt.git
    $ git checkout release-2.18.0
    $ git checkout -b 2.18
    $ git checkout origin/parsedown
    $ git checkout -b testing
    $ git rebase 2.18 # I know this maybe not a good example, but I want apply your changes to the 2.18.0 version and test them
    Instead of "composer install", due to my slow connection, I downloaded the official 2.18.0 tarball then extract and copy the "vendor" into the codes
    Perform a regular installation, turn on the Markdown Processing, create project and test issue
    The example codes are:
    example codes
    And the result:
    2.18.0 with default parsedown

3. It seems does not works with parsedown 1.8.0-beta-4 also:
    Download the version from https://github.com/erusev/parsedown/releases and replace the vendor/erusev/parsedown
    This result to the error:
    > PHP message: Call to undefined method Parsedown::blockFencedCodeComplete()\n/path/to/Parsedown.php: 303: MantisMarkdown - -> - blockFencedCodeComplete(DATA)\n/path/to/Parsedown.php: 52: Parsedown - -> - linesElements(DATA)

4. It works with parsedown 1.8.0-beta-7, but still have some issue:
    Please note some bug and note links followed with there number, but some links are not
    beta7

5. It seems broken the bug and note links when I turn off the Markdown Processing:
2.18.0 Markdown off

6. Please keep update the [#22180](https://mantisbt.org/bugs/view.php?id=22180), I was thought it is the parent of all Markdown issues, but when I got a problem, and look around and around, [#22175](https://mantisbt.org/bugs/view.php?id=22175) was closed(should be reopen and link to this I think), until I found [#24628](https://mantisbt.org/bugs/view.php?id=24628), [#23206](https://mantisbt.org/bugs/view.php?id=23206) and finally here.

7. Finally, is there any suggested test cases or complex Markdown example codes to test this issue? And thanks for all your work again!

**Comment by dregad** [comment]
@rimy thanks for your message and apologies for not replying sooner, I've been busy.

> 1. First of all, I am new to git and github, if I did anything wrong, please let me know, and I am happy to learn more. And I am not native English speaker, sorry for my English.

Don't hesitate to ask if you need help. Probably best to post your questions on Gitter.

> 3. It seems does not works with current parsedown with 2.18.0, here is my steps:

Parsedown 2.18 is in beta, and will not be released as stable since the developer is now focusing on a [2.0.0 version](https://github.com/erusev/parsedown/projects/1), as the changes are not backwards-compatible and would therefore break Semantic Versioning.

This is one of the reasons I left this PR aside (the others being the complexity of fixing all the use cases due to interaction with legacy, HTML tag based formatting, and lack of time)

> $ git rebase 2.18 # I know this maybe not a good example, but I want apply your changes to the 2.18.0 version and test them

I would advise to rebase onto _master_ or at least the latest version (2.21.0) and test that instead. It will reduce the risk of further merge conflicts.

> Instead of "composer install", due to my slow connection, I downloaded the official 2.18.0 tarball then extract and copy the "vendor" into the codes

A bit strange approach IMO, but I guess it should work.

> 4. It seems does not works with parsedown 1.8.0-beta-4 also:

Yes, as mentioned before 1.8.0 has backwards compatibility issues.

> 7. Please keep update the [#22180](https://mantisbt.org/bugs/view.php?id=22180), I was thought it is the parent of all Markdown issues, but when I got a problem, and look around and around, [#22175](https://mantisbt.org/bugs/view.php?id=22175) was closed(should be reopen and link to this I think), until I found [#24628](https://mantisbt.org/bugs/view.php?id=24628), [#23206](https://mantisbt.org/bugs/view.php?id=23206) and finally here.

Thanks for the heads up and sorry for the confusion. I've added some relationships between these issues to avoid other people getting lost...

> 8. Finally, is there any suggested test cases or complex Markdown example codes to test this issue? 

No, not yet. Creating a solid set of test cases is one of the desired improvements, as mentioned by @vboctor earlier. If you can contribute to that it would be great.

**Comment by dregad** [comment]
I have rebased the branch on top of latest master; there were several conflicts, which I hope were resolved correctly, but I do not have time to test. For reference, I kept a branch with the original version of the code (https://github.com/dregad/mantisbt/tree/parsedown-before-rebase).

**Comment by rimy** [comment]
> @rimy thanks for your message and apologies for not replying sooner, I've been busy.

@dregad thanks for your reply and that's ok, I am busy too :)

> > 1. It seems does not works with current parsedown with 2.18.0, here is my steps:
>
> Parsedown 2.18 is in beta, and will not be released as stable since the developer is now focusing on a [2.0.0 version](https://github.com/erusev/parsedown/projects/1), as the changes are not backwards-compatible and would therefore break Semantic Versioning.

Sorry for the confusion, I mean it not works with the parsedown version which currently used in MantisBT 2.18.0, and thanks for the information about parsedown 2.0.
 
> I would advise to rebase onto _master_ or at least the latest version (2.21.0) and test that instead. It will reduce the risk of further merge conflicts.

We are on 2.18.0 and not upgrade to latest yet, so I test the 2.18 first, and you are right I will test _master_, 2.21.0 and/or further versions, I will keep watching and reporting on this if I have time.

> No, not yet. Creating a solid set of test cases is one of the desired improvements, as mentioned by @vboctor earlier. If you can contribute to that it would be great.

I hope I could help, but more clarify please, how to write a test case for this? a phpt file? an SQL or bug import data? or just paste code here?

**Comment by nenadalm** [comment]
Since there seems to be no progress on this and the code seems to be unnecessarily complicated, we created alternative markdown plugin: https://github.com/Imatic-IT/imatic-mantis-formatting that uses https://github.com/thephpleague/commonmark.

**Comment by dregad** [comment]
> uses https://github.com/thephpleague/commonmark.

As mentioned in https://www.mantisbt.org/bugs/view.php?id=24810#c64266, please be aware that the CommonMark library requires PHP 7.2.5, while MantisBT's minimum requirement is 5.5.0 so the plugin may not be compatible with all MantisBT installations.

**Comment by TonyGravagno** [comment]
> Since there seems to be no progress on this and the code seems to be unnecessarily complicated, we created alternative markdown plugin: https://github.com/Imatic-IT/imatic-mantis-formatting that uses https://github.com/thephpleague/commonmark.

It's been noted elsewhere but I agree that it's time to more seriously consider moving this functionality to a completely separate plugin. Using Mantis in a few companies that do software development, we're constantly aware of the markdown issues.

I'm going to update my sites, disable default handling, add the ***Imatic*** plugin, and will do what I can to contribute to that, including back-porting detail noted in this thread. Honestly I'm not really qualified for this but we're dying out here. A new approach is required.

> the CommonMark library requires PHP 7.2.5, while MantisBT's minimum requirement is 5.5.0

PHP v8 is on the horizon. I do not see a good reason to hold back Mantis to v5 compatibility. There are some plugins that use deprecated functions and they will not port forward. Admins just need to make decisions about what functionality they want now, markdown versus 'something else', and configure according to priorities.

**Comment by nenadalm** [comment]
> I'm going to update my sites, disable default handling, add the Imatic plugin, and will do what I can to contribute to that

Cool. We just noticed there are some issues with margins around some texts caused by the plugin, just didn't get to it yet. I will investigate on some Monday (that's day I work on Mantis stuff currently among other things in our company). Also I didn't add the license gpl2 or later, which I probably should do on Monday.

> PHP v8 is on the horizon

I have no intentions supporting unsupported versions of php (https://www.php.net/supported-versions.php). Also I am currently working with 7.4 if possible. Supported older php version can be discussed **maybe** (though it's better if people fix problems with upgrading to the newer version - they will have to do it one day anyway), but unsupported php versions are **very unlikely** to be supported.

Should somebody decide to fork the plugin (or write new one) and support older php versions though, I will happily install theirs maintained plugin and stop developing ours :)

**Comment by jsandner** [comment]
> Since there seems to be no progress on this and the code seems to be unnecessarily complicated, we created alternative markdown plugin: https://github.com/Imatic-IT/imatic-mantis-formatting that uses https://github.com/thephpleague/commonmark.

Looks good and solves our problems with code blocks.

I had some trouble with installing commonmark by typing
`composer require league/commonmark`
it deleted *vendor/dapphp*, so I had to correct this manually.

Did you think about including syntax-highlighting in code-blocks?

**Comment by nenadalm** [comment]
> Did you think about including syntax-highlighting in code-blocks?

Nobody requested it and I don't use Mantis that much. Feel free to create ticket: https://github.com/Imatic-IT/imatic-mantis-formatting/issues/new

**Comment by ourtownamerica** [comment]
Thanks for your work on [Imatic Mantis Formatting](https://github.com/Imatic-IT/imatic-mantis-formatting), @nenadalm! Unfortunately the requirements didn't work for me, so I had to roll my own. Sharing here for visibility.

Here: [OT Mantis Markdown Fixer](https://github.com/ourtownamerica/OT-Mantis-Markdown-Fixer). Hope that helps someone.

**Comment by grummbeer** [comment]
fixed.

**Pull request closed without merging**